driftbase/flexmatch.py

```diff
@@ -0,0 +1,558 @@
+
+    connection_string = f"{ip_address}:{port}"
+    for player_id in players_to_notify:
+        event_data = {
+            "connection_string": connection_string,
+            "options": connection_info_by_player_id[player_id]
+        }
+        _post_matchmaking_event_to_members([player_id], "MatchmakingSuccess", event_data=event_data)
+
+def _process_matchmaking_cancelled_event(event):
+    for ticket in event["tickets"]:
+        ticket_id = ticket["ticketId"]
+        for player in ticket["players"]:
+            player_id = player["playerId"]
+            ticket_key = _get_player_ticket_key(player_id)
+            with _LockedTicket(ticket_key) as ticket_lock:
+                player_ticket = ticket_lock.ticket
+                if player_ticket is None:
+                    continue  # Normal, player cancelled the ticket and we've already cleared it from the cache
+                if player_ticket["Status"] == "COMPLETED" and ticket_id != player_ticket["TicketId"]:
+                    # This is not a flexmatch status, but I want to differentiate between statuses arising from the
+                    # cancelling of backfill tickets and other states
+                    player_ticket["Status"] = "MATCH_COMPLETE"
+                else:
+                    player_ticket["Status"] = "CANCELLED"
+                    _post_matchmaking_event_to_members([player_id], "MatchmakingCancelled")
+                ticket_lock.ticket = player_ticket
+
+def _process_accept_match_event(event):
+    game_session_info = event["gameSessionInfo"]
+    acceptance_by_player_id = {}
+    for player in game_session_info["players"]:
+        player_id = player["playerId"]
+        acceptance = player.get("accepted", None)
+        acceptance_by_player_id[player_id] = acceptance
+        if acceptance is not None:
+            with _LockedTicket(_get_player_ticket_key(int(player_id))) as ticket_lock:
+                player_ticket = ticket_lock.ticket
+                if player_ticket is None:
+                    log.error(f"Received acceptance event for player {player_id} who has no ticket.")
+                    return
+                if player_ticket["Status"] != "REQUIRES_ACCEPTANCE":
+                    log.error(f"Received acceptance event for player {player_id} who has a ticket in invalid state {player_ticket['Status']}.")
+                    return
+                for ticket_player in player_ticket["Players"]:
+                    if ticket_player['PlayerId'] == player_id:
+                        ticket_player["Accepted"] = acceptance
+                        ticket_lock.ticket = player_ticket
+                        break
+    _post_matchmaking_event_to_members(list(acceptance_by_player_id), "AcceptMatch", acceptance_by_player_id)
+
+def _process_accept_match_completed_event(event):
+    # This may be totally pointless as there should be a followup events to update the tickets to either 'searching', 'cancelled' or 'placing'
```

Comment:
Plural/singular mix makes this comment hard to understand; "as there should be *a* follow up *events*"

Reply:
Yes, in case of rejection, there should be as many follow up events as there were tickets/players.  If a potential match is rejected, those ticket holders who accepted will get a new 'searching' event, while those who rejected will get a 'failed' event. If all accepted, there will be a single success event for all of them.